doc/specifications/proposed/field_type.md

```diff
@@ -0,0 +1,308 @@
+    {
+        return $this->getName($value);
+    }
+
+    public function fromHash($hash)
+    {
+        if ($hash === null) {
+            return $this->getEmptyValue();
+        }
+
+        // The default constructor at least works for the top level objects.
+        // For more complex values a manual conversion is necessary.
+        return new {$this->valueClass}($hash);
+    }
+
+    public function toHash(SPIValue $value)
+    {
+        // Simplest way to ensure a deep structure is cloned and converted into
+        // scalars and has maps.
+        return json_decode(json_encode($value), true);
+    }
+
+Especially the hashing functions define assumptions about the vlaue
+object which might or might not be true. They expect the default
+constructor which assignes all array keys as properties on the value
+object. This must be documented when documenting custom value objects
+and their hashing methods. The hash methods can still be overwritten…
+
+This leaves only two methods to implement:
+
+- `getFieldTypeIdentifier()`
+
+  Obviously necessary.
+
+- `createValueFromInput($inputValue)`
+
+  This requires explanation how this integrates with the edit template
+  (which is missing entirely in the tuorial) and where this data
+  comes from. Web developers tend to try stuff in the web interface
+  and the connecting dots are missing to be able to do this.
+
+  We *could* also make this method optional by adding a
+  `::fromInput()` method to the value object and providing default
+  implementations for the pre-defined value classes (see below).
+
+### The Type Class
+
+Right now it is quite tedious to develop a new field type because of
+various issues:
+
+- Development Experience: Watching fatal errors and SF HTML error
+  responses in AJAX requests in the browser debug bar. (Probably
+  resolved in next version)
+
+- Tutorial missing edit and settings templates – the latter causes
+  Internal Server Errors and a broken view.
+
+  There is a pull request to make the settings template optional again
+  and thus it should be documented in the relevant tutorial section
+  (about settings).
+
+  The edit template seems necessary to properly try out the field type
+  and should be documented.
+
+- Field type indentifier MUST NOT contain dashes so that Twig template
+  blocks work. We should probably document (and verify or normalize)
+  them as `/^[a-zA-Z0-9_]+$/`.
+
+  Internal normalization probably isn't an option since a developer
+  would have to use different identifiers in Twig template blocks then
+  the one specified in the field type implementation. We could assert
+  on a proper identifier in the test case. The documentation suggests
+  to use `.` as a separator in the name – we should ensure this works
+  fine with Twig since I guess this could be considered an operator by
+  Twig, too.
+
+We should probably document the field type base test class
+`eZ\Publish\Core\FieldType\Tests\FieldTypeTest` which allows to test
```

Comment:
The problem with `Core` namespace is that we tend to change its "contents" without a notice. If we intend for developers to extend it, maybe we should move it to `SPI`?